lib/TPP/Dialect/Tpp/TppUtils.cpp

```diff
@@ -314,15 +314,55 @@ static bool allIndexingsAreProjectedPermutation(linalg::GenericOp genericOp) {
   });
 }
 
-// Return true if the linalg.generic can be mapped to a tpp.add.
-bool isTppAdd(linalg::GenericOp linalgOp) {
+// Only parallel loops at buffer semantics with static shapes.
+bool hasMappingToTppConditions(linalg::GenericOp linalgOp) {
   if (linalgOp.getNumLoops() != linalgOp.getNumParallelLoops())
     return false;
-  if (linalgOp.getNumDpsInits() != 1)
-    return false;
   if (linalgOp.hasTensorSemantics())
     return false;
-  if (!hasStaticShape(linalgOp))
+  return hasStaticShape(linalgOp);
+}
+
+static bool hasOneUser(Value val) {
+  return std::distance(val.getUsers().begin(), val.getUsers().end()) == 1;
+}
+
+static bool hasZeroUser(Value val) {
+  return std::distance(val.getUsers().begin(), val.getUsers().end()) == 0;
+}
+
+// Return true if the operation is binary.
+static bool isBinaryOp(linalg::GenericOp linalgOp) {
+  if ((linalgOp.getNumDpsInputs() == 1) && (linalgOp.getNumDpsInits() == 1)) {
+    Value outArg = linalgOp.getRegionOutputArgs()[0];
```

Comment:
If the region (of a non-binary linalg region) has no output arguments, this will crash. You only check for DPS inputs in the if expression.

Perhaps it could be easier to just put a simple check outside of both IFs in the first line:
```
  if (linalgOp.getOutputs().size() < 1)
    return false;
  Value outArg = linalgOp.getRegionOutputArgs()[0];
  ...
```

Reply:
I think, at least in the generic, getNumDpsInits is the number of output: https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L221

Reply:
https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/include/mlir/Interfaces/DestinationStyleOpInterface.td#L78

Reply:
Yeah, it looks like inits map to output - I found this example a bit clearer: https://github.com/llvm/llvm-project/blob/4c6d6fc176a115d196194c6d5091d3279d27fffa/mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp#L558

Although, I wouldn't mind some more obvious API for that.

Reply:
Welcome to MLIR